LibTrixi.jl/test/Project.toml

```diff
@@ -0,0 +1,23 @@
+[deps]
+OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
+Trixi = "a7f1ee26-1774-49b1-8366-f1abc58fbfcb"
+MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
```

Comment:
```suggestion
```

I think these are not needed (and neither are their compat entries). IIRC, you only need to add packages here that you use in `runtest.jl` (and its includes) - you do not need this for packages that are used by LibTrixi.jl, since they will already be installed automatically with LibTrixi.jl.

Please verify that I did not miss anything, but AFAICT, your `[deps]` should only contain the entry for `Test` and the `[compat]` section can be removed entirely.

Reply:
Thank you for the explanation! To be honest, I did not fully understand the mechanism.

I tried to cleanup `test/Project.toml` now.
With only `Test` in it, I can run the tests successfully via
```
~/libtrixi/LibTrixi.jl $ JULIA_DEPOT_PATH=~/install/libtrixi-julia/julia-depot LIBTRIXI_DEBUG=all julia --project=./ test/runtests.jl
```
Via julia
```
~/libtrixi/LibTrixi.jl $ JULIA_DEPOT_PATH=~/install/libtrixi-julia/julia-depot LIBTRIXI_DEBUG=all julia --project=./ test/runtests.jl

julia> import Pkg

julia> Pkg.test()
```
however, it does not work. It is missing `Trixi` and tells me to `Pkg.add` it.

With `Test`, `Trixi`, and `OrdinaryDiffEq` in Project.toml it works again.

Reply:
Can you please fix the Julia command for the second case? There seems to be a copy paste error...

Reply:
Indeed!

Reply:
I think if you just run `Pkg.test()`, the active environment should be the folder LibTrixi.jl. Is this the case? Also, what is the exact error? I don't understand why it would miss "Trixi", since `using Trixi` is never executed in the files located in `test/`

Reply:
Another idea:
`test/runtests.jl` includes `test/test_interface.jl`, which calls `trixi_initialize_simulation(libelixir)`, which includes `libelixir_tree1d_dgsem_advection_basic.jl` (via `Base.include(Main, libelixir)`), which starts with
```
using LibTrixi
using Trixi
using OrdinaryDiffEq
``` 
Could this be the reason?

Reply:
Hm. You might be right!

Is there a way to easily make the module configurable, i.e., such that it is `Main` by default (especially when called from C), but to allow one to use a different module from Julia? E.g., a second version of `trixi_initialize_simulation` with an additional first parameter `module`, and then connect them together similar to this:
```julia
function trixi_initialize_simulation(module, libelixir::String)
    ....
end
function trixi_initialize_simulation(libelixir::String)
    trixi_initialize_simulation(Main, libelixir)
end
```
?

Reply:
Certainly! This could also be a parameter with default value, couldn't it?

But I do not get the point. Would I then used `Module=LibTrixi` when testing?

Reply:
> Certainly! This could also be a parameter with default value, couldn't it?

Yes. However, this is how we do it in [Trixi.jl](https://github.com/trixi-framework/Trixi.jl/blob/e22b11e5859b5168e24dabb4cb34a6a35fce3e66/src/auxiliary/special_elixirs.jl#L38C24-L48) and how, I believe, is custom in other Julia base functions.

> Would I then used Module=LibTrixi when testing?

No, you'd use something like `@__MODULE__` which gives the currently active module. The only issue I can with this is that it will not allow us to fully test the C interface from within Julia anymore, since we cannot pass an object of type `Module` through the C interface. But maybe that's OK?

Reply:
I'm afraid this is what I am currently doing. 
https://github.com/trixi-framework/libtrixi/blob/35485c1370437a69e7e51747447c6a6361355459/LibTrixi.jl/test/test_interface.jl#L23
But of course the test could be rewritten.

I assume `@__MODULE__` cannot just always be used?

Reply:
The same as before
```
*** Base.include with module Main.TestInterface
libtrixi tests: Error During Test at /home/bene/libtrixi_main/LibTrixi.jl/test/runtests.jl:3
  Got exception outside of a @test
  LoadError: ArgumentError: Package Trixi not found in current path.
  - Run `import Pkg; Pkg.add("Trixi")` to install the Trixi package.
```

Reply:
Can you also give me the full command you use to trigger this error with the current branch?

I don't think that evaluating this inside the module `LibTrixi` is the right choice, which is why I am hesitant...

Reply:
Sure! I just pushed the experimental version.

```
~/libtrixi_main/LibTrixi.jl (bg/julia-test *+) $ LIBTRIXI_DEBUG=all julia --project=./
"]"
(LibTrixi) pkg> test
```

Don't get me wrong! I am not claiming that it would be right to do it in another way. I just have no clue.  ;)

Reply:
In this command above, you do not set the `JULIA_DEPOT_PATH`, are you sure about that?

Reply:
I'm not really sure about the implications. So I actually tried both, with and without.

Reply:
I think I get it now, and you were right in the first place:

Since we *do* evaluate the elixirs in `Main`, the packages that are `include`d in the libelixirs actually do need to be loaded in the project that is currently active. I was wrong in my analysis back then.

Now, the question is, how to proceed here? I am not 100% sure that a *canonical* approach would look like, thus I asked on [Discourse](https://discourse.julialang.org/t/how-to-canonically-load-and-run-user-provided-functions-from-a-file/103180) in the hope that it will shed some light on this. In the meantime, I see three options for moving forward:
1. Add back the dependencies to the `Project.toml` (including the same compat bounds)
2. Load the libelixirs in the `LibTrixi` module
3. Remove `using Trixi, OrdinaryDiffEq` from the libelixirs.

I think options 2) and 3) are too restrictive, since they do not allow users to provide libelixirs with additional dependencies (maybe someone would like to use code from another third-party package in the future?). Thus I suggest to use 1) and maybe add a comment to the compat sections in `Project.toml` and `test/Project.toml`, stating that if compat bounds are adapted, it should be verified that they match the ones in the other file for packages that are present in both files.

Reply:
Thanks for you ongoing efforts!

I would certainly be fine with 1!

But, out if interest, what is the downside of 2?
If a user would like to use another Package, she would need to add it to LibTrixi's Project.toml, and not our `libtrixi-julia` project, right?
(plus issues with overwriting you mentioned on Discourse)

Reply:
> But, out if interest, what is the downside of 2?
If a user would like to use another Package, she would need to add it to LibTrixi's Project.toml, and not our `libtrixi-julia` project, right?

Correct. And there's the problem: We don't want people to have to modify LibTrixi.jl just to use a different package - that would be imho very non-Julian and in general very inconvenient.

But we probably want to add it to our documention that any package that one wants to use in an libelixir needs to be added *a priori* to the `libtrixi-julia` project.

Reply:
I agree. Both are important points!